File: swift/Sources/AlloyTime/TimeMachine.swift
import Foundation
import Observation

/// Key-value store the TimeMachine persists into. UserDefaults conforms; tests
/// inject a dictionary. Mirrors the TS `TimeMachineStorage`.
public protocol TimeMachineStorage: AnyObject {
    func getItem(_ key: String) -> String?
    func setItem(_ key: String, _ value: String)
    func removeItem(_ key: String)
}

extension UserDefaults: TimeMachineStorage {
    public func getItem(_ key: String) -> String? { string(forKey: key) }
    public func setItem(_ key: String, _ value: String) { set(value, forKey: key) }
    public func removeItem(_ key: String) { removeObject(forKey: key) }
}

/// Time Machine model: an optional frozen instant + optional zone override that
/// every Ally face can read instead of the real clock. Pure model — apps layer
/// their own ticking/reactivity on top. Mirrored twin of `time-machine.ts`.
@Observable
public final class TimeMachine {
    public private(set) var mock: Date?
    public private(set) var mockTimeZone: String?

    private let localZone: String
    private let storage: TimeMachineStorage?
    private let mockKey: String
    private let tzKey: String
    nonisolated(unsafe) private static let iso = ISO8601DateFormatter()

    public init(localZone: String = TimeZone.current.identifier,
                storage: TimeMachineStorage? = UserDefaults.standard,
                namespace: String = "ally",
                isUsableZone: ((String) -> Bool)? = nil)
    {
        self.localZone = localZone
        self.storage = storage
        mockKey = "\(namespace).clock.mock"
        tzKey = "\(namespace).clock.tz"
        let usable = isUsableZone ?? { ZoneCatalog.resolve($0) != nil }

        if let stored = storage?.getItem(mockKey),
           let date = Self.parseISO(stored) { mock = date }
        if let zone = storage?.getItem(tzKey), usable(zone) { mockTimeZone = zone }
    }

    public var isMocked: Bool { mock != nil || mockTimeZone != nil }
    public func now(_ realNow: Date) -> Date { mock ?? realNow }
    public func timeZone() -> String { mockTimeZone ?? localZone }

    public func setMock(_ date: Date) {
        mock = date
        storage?.setItem(mockKey, Self.iso.string(from: date))
    }

    public func clearMock() {
        mock = nil
        storage?.removeItem(mockKey)
    }

    /// Selecting the device's local zone is "follow local", not a mock.
    public func setTimeZone(_ zone: String) {
        if zone == localZone {
            clearTimeZone()
            return
        }
        mockTimeZone = zone
        storage?.setItem(tzKey, zone)
    }

    public func clearTimeZone() {
        mockTimeZone = nil
        storage?.removeItem(tzKey)
    }

    /// ISO 8601 with or without fractional seconds (JS `toISOString()` includes
    /// milliseconds; the plain formatter does not parse them).
    private static func parseISO(_ s: String) -> Date? {
        if let d = iso.date(from: s) { return d }
        let fractional = ISO8601DateFormatter()
        fractional.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
        return fractional.date(from: s)
    }
}

File: web/packages/alloy-time/src/time-machine.ts
// Time Machine model: an optional frozen instant + optional zone override that
// every Ally face can read instead of the real clock. Pure model — the app
// wraps it in its own reactive layer (Angular signals / SwiftUI Observation).
// Mirrored twin of AlloyTime's TimeMachine.

export interface TimeMachineStorage {
  getItem(key: string): string | null;
  setItem(key: string, value: string): void;
  removeItem(key: string): void;
}

export interface TimeMachineOptions {
  localZone: string;
  storage?: TimeMachineStorage | null;
  namespace?: string;
  isUsableZone?: (id: string) => boolean;
}

// A zone is usable only if Intl accepts it (stale persisted values are ignored).
function intlAcceptsZone(id: string): boolean {
  try {
    new Intl.DateTimeFormat('en-US', { timeZone: id });
    return true;
  } catch {
    return false;
  }
}

export class TimeMachine {
  private readonly localZone: string;
  private readonly storage: TimeMachineStorage | null;
  private readonly mockKey: string;
  private readonly tzKey: string;
  private mockNow: Date | null = null;
  private mockZone: string | null = null;

  constructor(options: TimeMachineOptions) {
    this.localZone = options.localZone;
    this.storage = options.storage ?? null;
    const ns = options.namespace ?? 'ally';
    this.mockKey = `${ns}.clock.mock`;
    this.tzKey = `${ns}.clock.tz`;
    const usable = options.isUsableZone ?? intlAcceptsZone;

    const storedMock = this.read(this.mockKey);
    if (storedMock) {
      const date = new Date(storedMock);
      if (!isNaN(date.getTime())) this.mockNow = date;
    }
    const storedZone = this.read(this.tzKey);
    if (storedZone && usable(storedZone)) this.mockZone = storedZone;
  }

  get mock(): Date | null { return this.mockNow; }
  get mockTimeZone(): string | null { return this.mockZone; }
  get isMocked(): boolean { return this.mockNow !== null || this.mockZone !== null; }

  now(realNow: Date): Date { return this.mockNow ?? realNow; }
  timeZone(): string { return this.mockZone ?? this.localZone; }

  setMock(date: Date): void {
    this.mockNow = date;
    this.write(this.mockKey, date.toISOString());
  }

  clearMock(): void {
    this.mockNow = null;
    this.remove(this.mockKey);
  }

  // Selecting the device's local zone is "follow local", not a mock.
  setTimeZone(zone: string): void {
    if (zone === this.localZone) {
      this.clearTimeZone();
      return;
    }
    this.mockZone = zone;
    this.write(this.tzKey, zone);
  }

  clearTimeZone(): void {
    this.mockZone = null;
    this.remove(this.tzKey);
  }

  // Storage may be absent or throwing (private browsing); in-memory state wins.
  private read(key: string): string | null {
    try { return this.storage?.getItem(key) ?? null; } catch { return null; }
  }
  private write(key: string, value: string): void {
    try { this.storage?.setItem(key, value); } catch { /* keep in-memory */ }
  }
  private remove(key: string): void {
    try { this.storage?.removeItem(key); } catch { /* keep in-memory */ }
  }
}
